portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -28,13 +35,61 @@
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    ArrayList<String> messages = new ArrayList<String>();
-    messages.add("first message");
-    messages.add("second message");
-    messages.add("third message");
-    Gson gson = new Gson();
-    String jsonMessage = gson.toJson(messages);
+    ArrayList<Message> messages = new ArrayList<Message>();
+
+    Query query = new Query("Message");
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()) {
+      String email = (String) entity.getProperty("email");
+      String name = (String) entity.getProperty("name");
+      String message = (String) entity.getProperty("message");
+
+      Message messageObj = new Message(name, email, message);
+      messages.add(messageObj);
+    }
+
+    String jsonMessage = convertToJson(messages);
     response.setContentType("application/json;");
     response.getWriter().println(jsonMessage);
   }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String email = getParameter(request, "email");
+    String name = getParameter(request, "name");
+    String message = getParameter(request, "message");
+
+    if (email.isEmpty() || name.isEmpty() || message.isEmpty()) {
+      response.setContentType("text/html;");
+      response.getWriter().println("<meta http-equiv='refresh' content='3;URL=index.html'>");
+      response.getWriter().println("<p style='color:red;'>All form fields must not be empty</p>");
+      return;
+    }
+
+    Entity newEntity = new Entity("Message");
+    newEntity.setProperty("email", email);
+    newEntity.setProperty("name", name);
+    newEntity.setProperty("message", message);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(newEntity);
+
+    response.sendRedirect("/index.html");
+  }
+
+  private String getParameter(HttpServletRequest request, String name) {
+    String value = request.getParameter(name);
+    if (value == null || value.isEmpty()) {
```

Comment:
|| value.isEmpty() is unnecessary. 

I think over here, it's better to throw an exception here for null value rather than silently covering it up, since it's something that shouldn't happen, and may indicate a bug.

Reply:
then what if the input is an empty string?

Reply:
If it's empty string, it will just be returned, so the condition is unnecessary. 

As for whether you want to throw an error, I guess it depends if you want to throw the same error for empty string and null.

Reply:
i don't understand why null should not happen, what if the client is not a browser and he specify the value to be null in json?

Reply:
There is already an assumption that it's a browser from the responses returned (redirect, html, etc.). But regardless of the client, these fields are treated as compulsory fields, so on the server, we expect the client to fill it in. if compulsory fields are missing, then the assumption is that the client is not constructing the request correctly. This is slightly different from the case where the field exists but is empty, which may be a user input problem.